File: Cpp/src/hashing/ransom_note.c
#include <stdbool.h>

int get_char_index(char character) {
    return character - 'a';
}

bool can_construct(const char *ransomNote, const char *magazine) {
    int hash_table[26] = {0};

    for (int i = 0; magazine[i] != '\0'; i++) {
        int index = get_char_index(magazine[i]);
        hash_table[index]++;
    }
    for (int i = 0; ransomNote[i] != '\0'; i++) {
        int index = get_char_index(ransomNote[i]);
        hash_table[index]--;
        if (hash_table[index] < 0) {
            return false;
        }
    }
    return true;
}

File: Python/src/hashing/ransom_note.py
from typing import Dict


class RansomNote:

    def can_construct(self, ransom_note: str, magazine: str) -> bool:
        magazine_map: Dict[str, int] = dict()

        for character in magazine:
            frequency = magazine_map.get(character, 0)
            magazine_map[character] = frequency + 1

        for character in ransom_note:
            frequency = magazine_map.get(character, 0)
            magazine_map[character] = frequency - 1
            if magazine_map[character] < 0:
                return False

        return True
